### PR TITLE
add synthesis plugin docs

### LINKED ISSUE #11675: Add API docs section for synthesis plugins post-1.0.0rc1
### What should we add?

We should have a dedicated page in the api docs that enumerate all the built-in transpiler synthesis plugins. These should list the plugin name and the details of all the plugins. Right now we only have this page:

https://github.com/Qiskit/qiskit/blob/main/docs/apidoc/transpiler_builtin_plugins.rst

which only lists aqc. 

This was attempted in #10657 but was removed from that PR because it was larger scope than just adding that method. What was attempted in that PR was to use the class docstrings for the plugin classes. The issue with that though is that the plugin method interface isn't actually the class, but instead is the plugin name. It was a bit awkward to link to all the classes in the documentation but say the interface being documented isn't the thing you should use, that link just contains the details of some other interface not explicitly documented.

We should try to include this in the final 1.0.0 release, as it's only documentation changes we can do this post-rc1 as it won't change any interfaces being released.

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

Adds API docs section for synthesis plugins.
Fixes #11675.

UPDATE:

For each relevant high-level-object (e.g., `Clifford`, `LinearFunction` or `PermutationGate`) we are adding a table comparing different plugins available for this object, and the links to the available plugins (to easy finding the relevant documentation).

### CODE REVIEW

File: qiskit/transpiler/passes/synthesis/plugin.py [hunk 1]
```diff
@@ -336,6 +336,161 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
...
+
+
+Unitary Synthesis Plugins
+-------------------------
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   /qiskit/transpiler/passes/synthesis/unitary_synthesis/DefaultUnitarySynthesis
+   /qiskit/transpiler/passes/synthesis/solovay_kitaev_synthesis/SolovayKitaevSynthesis
```
Comment (ShellyGarion) [inline qiskit/transpiler/passes/synthesis/plugin.py]: I think that `AQCSynthesisPlugin` also belongs here:
https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/synthesis/aqc_plugin.py

Comment (ShellyGarion) [reply qiskit/transpiler/passes/synthesis/plugin.py]: should these docs replace the docs here: https://docs.quantum.ibm.com/api/qiskit/dev/transpiler_passes#synthesis
as it seems that there is some overlap between them

**Comment by ShellyGarion** [comment]
I don't think that the file `docs/apidoc/synthesis_aqc.rst` should be removed. It contains the entire AQC synthesis code, and not the `AQCSynthesisPlugin`, so now the following link from `qiskit.synthesis` does not work:
`The Approximate Quantum Compiler is available here: qiskit.synthesis.unitary.aqc`

**Comment by ElePT** [comment]
> I don't think that the file `docs/apidoc/synthesis_aqc.rst` should be removed. It contains the entire AQC synthesis code, and not the `AQCSynthesisPlugin`, so now the following link from `qiskit.synthesis` does not work: `The Approximate Quantum Compiler is available here: qiskit.synthesis.unitary.aqc`

@ShellyGarion Let me take a look, I removed the file because sphinx was complaining that it wasn't part of any toctree (because the builtin plugin file was the only one using it), but we should be able to make the link work without having an individual file for it. If I can't find the way, let's bring it back. At least, do you think the synthesis plugin page looks like it was intended to?

**Comment by ShellyGarion** [comment]
I think that the `AQCSynthesisPlugin` looks OK, but this page has disappeared:
https://docs.quantum.ibm.com/api/qiskit/synthesis_aqc

**Comment by ElePT** [comment]
@ShellyGarion I think that [cf11cb7](https://github.com/Qiskit/qiskit/pull/11763/commits/cf11cb740e3f136c4440df459dce32dfc7a247e0) has fixed the AQC page. I managed to generate it without an additional rst file.

**Comment by ShellyGarion** [comment]
> @ShellyGarion I think that [cf11cb7](https://github.com/Qiskit/qiskit/pull/11763/commits/cf11cb740e3f136c4440df459dce32dfc7a247e0) has fixed the AQC page. I managed to generate it without an additional rst file.

Indeed, it looks much better now. thanks!

image

File: qiskit/transpiler/passes/synthesis/aqc_plugin.py [hunk 2]
```diff
@@ -10,7 +10,14 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 """
-An AQC synthesis plugin to Qiskit's transpiler.
+==============================================================================
+AQC Synthesis Plugin (in :mod:`qiskit.transpiler.passes.synthesis.aqc_plugin`)
+==============================================================================
```
Comment (ElePT) [inline qiskit/transpiler/passes/synthesis/aqc_plugin.py:15]: I tried removing these titles for the aqc/solovay-kitaev/unitary synthesis plugins but I think the result looks a bit strange:

Screenshot 2024-02-14 at 15 46 08

So I decided to leave the titles, but either way is possible:

Screenshot 2024-02-14 at 15 47 56

File: qiskit/transpiler/passes/synthesis/high_level_synthesis.py [hunk 3]
```diff
@@ -11,7 +11,127 @@
...
+      - all-to-all
+      - same as ``"basic"``
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   BasicSynthesisPermutation
+   ACGSynthesisPermutation
+   KMSSynthesisPermutation
+   TokenSwapperSynthesisPermutation
```
Comment (ElePT) [inline qiskit/transpiler/passes/synthesis/high_level_synthesis.py]: I also tried integrating these with little success, but I think that having 2 tables is not that bad

**Comment by jakelishman** [comment]
This is marked as closing an issue in the 1.0.0 milestone. Is it ready to review/merge now?

We can always tweak documentation after the release, but it'd be good to have at least the main skeleton in to begin with (and this already looks fairly done?). We could merge this and have more PRs later if that's convenient?

**Comment by alexanderivrii** [comment]
Sorry for the delay. I would like to make one more quick pass over the text in the tables. And many thanks to @ElePT for turning the initial mess of this PR to the current form.

**Review by jakelishman — APPROVED** [review]
Thanks for this, Sasha, Elena and Shelly - it looks like a really big improvement over what we've (not!) got so far.

**Review by ShellyGarion — APPROVED** [review]

**Pull request merged**